### ISSUE TITLE
Memory leaks 

### ISSUE THREAD

**Issue by FrankHogervorst** [Tudat/tudat#206]
As discussed in the meeting with @DominicDirkx, here is the output of valgrind and gdb for my pagmo problem.  I inserted two gdb outputs since they gave slightly different output for two consecutive runs.

The output when trying to perform parallel computing:
[gdb_parallelComputedPagmo.txt](https://github.com/Tudat/tudat/files/1105132/gdb_parallelComputedPagmo.txt)
[gdb_parallelComputedPagmo_run2.txt](https://github.com/Tudat/tudat/files/1105135/gdb_parallelComputedPagmo_run2.txt)
[valgrind_parallelComputedPagmo_fullLeakCheck.txt](https://github.com/Tudat/tudat/files/1105133/valgrind_parallelComputedPagmo_fullLeakCheck.txt)

The output without parallel computing:
[gdb_pagmo.txt](https://github.com/Tudat/tudat/files/1105136/gdb_pagmo.txt)
[valgrind_pagmo_fullLeakCheck.txt](https://github.com/Tudat/tudat/files/1105137/valgrind_pagmo_fullLeakCheck.txt)

**Comment by DominicDirkx** [comment]
As we discussed, the multi-core issue is most likely due to teh fact that each thread would need its own Body object, which is not how things are set up now.

The memory leaks in the AerodynamicAngleCalculator are, I think, because this class has no destructor. Could you add one (empty), and try it again. To figure out the issue with the LagrangeInterpolatorm could you add a cout to the 

LagrangeInterpolator constructors
LagrangeInterpolator destructor
TabulatedCartesianEphemeris destructor

Also, could you add a destructor to the Body class, and add a cout message to both the constructor and destructor of the Body? Running a single iteration of the algorithm should generate sufficient output, and will probably make it easier to interpret :)

Let me know what comes out,

**Comment by FrankHogervorst** [comment]
Adding a destructor for the AerodynamicAngleCalculator gave the same result:
[valgrind_pagmo_fullLeakCheck_AerodynamicAngleCalculatorDestructor.txt](https://github.com/Tudat/tudat/files/1124608/valgrind_pagmo_fullLeakCheck_AerodynamicAngleCalculatorDestructor.txt)

The valgrind output with all the cout is shown here:
[valgrind_pagmo_fullLeakCheck_with_cout_and_destructors.txt](https://github.com/Tudat/tudat/files/1124637/valgrind_pagmo_fullLeakCheck_with_cout_and_destructors.txt)

**Comment by DominicDirkx** [comment]
So, the problem with the Body objects, and as a result the rest, seems to be that 6 bodies are created, but only 2 are de-allocated. I think this a result of cross-dependencies, which means that the share count of the pointers never reaches zero. I'll see if I can write a fix for this, which would clear the contents of the member objects.

**Comment by DominicDirkx** [comment]
I've written a small piece of code that manually clears the Lagrange Interpolator objects in the ephemerides, and have pushed it to this branch:

https://github.com/DominicDirkx/tudat/tree/MemoryLeakMitigation

Merge this branch into your code. Make sure to commit any new code you have first so that you can easily reverse the merge if needed.

You can have a look at the new code here:

https://github.com/DominicDirkx/tudat/commit/c6913de9cf981fe70244ed832ffab5d578eefe8b

In short, after you are done with the bodyMap, you can call the (somewhat optimistically named) function:

emptyBodyMapMemory( bodyMap )

Let me know if using this works for you.

**Comment by FrankHogervorst** [comment]
I called emptyBodyMapMemory( bodyMap)  at the end of my code, this gave similar results as without.
[valgrind_pagmo_fullLeakCheck_withemptyBodyMapMemory_function.txt](https://github.com/Tudat/tudat/files/1135005/valgrind_pagmo_fullLeakCheck_withemptyBodyMapMemory_function.txt) 
[valgrind_pagmo_fullLeakCheck_withOUTemptyBodyMapMemory_function.txt](https://github.com/Tudat/tudat/files/1135040/valgrind_pagmo_fullLeakCheck_withOUTemptyBodyMapMemory_function.txt)

**Comment by DominicDirkx** [comment]
Hmmm... Could you put a cout statement in the new 'clear' functions in both the OneDimensionalInterpolator and the LagrangeInterpolator? 

At the end of the clear statements, could you cout the sizes of the member vectors that should be cleared? They should be zero, but apparently they're not....

**Comment by FrankHogervorst** [comment]
I accidentally uplouded the wrong version for the run without the clearing. Here is the correct one:
[valgrind_pagmo_fullLeakCheck_withOUTemptyBodyMapMemory_function.txt](https://github.com/Tudat/tudat/files/1135080/valgrind_pagmo_fullLeakCheck_withOUTemptyBodyMapMemory_function.txt)

**Comment by FrankHogervorst** [comment]
My previous statement that it gave similar results was not true. It fixed the loss of 47,999,856 bytes. But the loss of 50,331,648 bytes is still present. 

The couts of the clear statements:

Progress 1/1
Size of members OneDimensionalInterpolator
0
0
Size of members OneDimensionalInterpolator
0
0
Size of members LagrangeInterpolator
0
0
Size of members OneDimensionalInterpolator
0
0
Size of members OneDimensionalInterpolator
0
0
Size of members OneDimensionalInterpolator
0
0
Size of members LagrangeInterpolator
0
0
Size of members OneDimensionalInterpolator
0
0
cleared BodyMap memory 
Finished in 0.099069 seconds [CPU Clock] 
Awesome this thing completed a full run!
Body destructor called
TabulatedCartesianEphemeris destructor called
LagrangeInterpolator destructor called
/home/frank/tudatBundle/tudatApplications/thesis/bin/applications/application_pagmoTest exited with code 0

**Comment by DominicDirkx** [comment]
Hi Frank, Could you post the full output of the program?

**Comment by FrankHogervorst** [comment]
Updated the previous post.

**Comment by DominicDirkx** [comment]
It looks like 2 Lagrange Interpolators are created, and 2 are emptied, so it's odd that one of the errors persists.

What/where exactly have you placed your new cout statements (in teh interpolators). Could you add explicitly to the statements what they denote, and in which function/class?

Also, could you ptint the 'this' pointer in both the constructor, destructor and clear functions of the interpolators?

**Comment by FrankHogervorst** [comment]
The couts are added at the end of the clear statements. I've added some extra information to the cout, let me know if it gives enough information:

```

Starting /home/frank/tudatBundle/tudatApplications/thesis/bin/applications/application_pagmoTest...
Test case selected
Create pagmo problem
OrbitSimulator default constructor called
OrbitSimulator constructor called
Body  constructor called
Body  constructor called
LagrangeInterpolator constructor from map of independent/dependent data called
Corresponding pointer: 0x154d5f30
LagrangeInterpolator constructor from map of independent/dependent data called
Corresponding pointer: 0x1b8644e0
Body  constructor called
Low thrust transfer problem created
Algorithm selected
Population created
Starting evolve loop ...
Progress 1/1
Function clearInterpolator called inside oneDimensionalInterpolator.h,
The independentValues_ parameter has size:0
the dependentValues_ parameter has size:0
Corresponding pointer to the onedimensional interpolater object: 0x21bf1ae0
Function clearInterpolator called inside oneDimensionalInterpolator.h,
The independentValues_ parameter has size:0
the dependentValues_ parameter has size:0
Corresponding pointer to the onedimensional interpolater object: 0x21bf20d0
Function clearDerivedInterpolator called inside lagrangeInterpolator.h,
The denominators parameter has size:0
the independentVariableDifferenceCache parameter has size:0
Corresponding pointer to the cleared lagrange interpolater object: 0x1b8644e0
Function clearInterpolator called inside oneDimensionalInterpolator.h,
The independentValues_ parameter has size:0
the dependentValues_ parameter has size:0
Corresponding pointer to the onedimensional interpolater object: 0x1b8644e0
Function clearInterpolator called inside oneDimensionalInterpolator.h,
The independentValues_ parameter has size:0
the dependentValues_ parameter has size:0
Corresponding pointer to the onedimensional interpolater object: 0x1b863ae0
Function clearInterpolator called inside oneDimensionalInterpolator.h,
The independentValues_ parameter has size:0
the dependentValues_ parameter has size:0
Corresponding pointer to the onedimensional interpolater object: 0x1b8640f0
Function clearDerivedInterpolator called inside lagrangeInterpolator.h,
The denominators parameter has size:0
the independentVariableDifferenceCache parameter has size:0
Corresponding pointer to the cleared lagrange interpolater object: 0x154d5f30
Function clearInterpolator called inside oneDimensionalInterpolator.h,
The independentValues_ parameter has size:0
the dependentValues_ parameter has size:0
Corresponding pointer to the onedimensional interpolater object: 0x154d5f30
cleared BodyMap memory 
Finished in 0.09878 seconds [CPU Clock] 
Awesome this thing completed a full run!
Body destructor called
TabulatedCartesianEphemeris destructor called
LagrangeInterpolator destructor called
Corresponding pointer: 0x1b8644e0
OneDimensionalInterpolator destructor called
Corresponding pointer: 0x21bf20d0
OneDimensionalInterpolator destructor called
Corresponding pointer: 0x21bf1ae0
OneDimensionalInterpolator destructor called
Corresponding pointer: 0x1b8644e0

```

**Comment by DominicDirkx** [comment]
Very strange. Could you modify your compiler.cmake so that no compiler optimization is performed?

There are a bunch of -Ox settings in this file. If you change them to -O0, clean and recompile the code, it will run without compiler optimizations. This may give some more insight into where the issue is actually occuring. The strange this is that a leak in a vector< double > has now been fixed, but an issue in a vector< Vector6d > (presumably dependent variables or independentVariableDifferenceCache) and a vector< vector< double > > (presumably denominators) persist....

**Comment by FrankHogervorst** [comment]
Where can I find the compiler cmake?

**Comment by DominicDirkx** [comment]
tudatBundle/external/cmake/

**Comment by FrankHogervorst** [comment]
I recompiled with all -Ox settings set to -Oo. Valgrind output:
[valgrind_pagmo_fullLeakCheck_compilersettings-O0.txt](https://github.com/Tudat/tudat/files/1136074/valgrind_pagmo_fullLeakCheck_compilersettings-O0.txt)

**Comment by DominicDirkx** [comment]
Ok, so it looks like that the problem lies with the a vector of doubles and a vector of Vector6ds, both of which are allocated through a 'push_back' command. So, it seems that the problem is that the independentValues_ and dependentValues_ are not cleared properly. I don't see why, because:

     independentValues_.clear( );
     dependentValues_.clear( );

Is clearly in the clearInterpolator function in OneDimensionalInterpolator..... 

Could you try moving these 2 clear commands into the clearDerivedInterpolator function of LagrangeInterpolator as well? I really don't see why it should matter, but then again, I also don't see why the memory isn't cleared to begin with.

**Comment by FrankHogervorst** [comment]
Nope that doesn't help either. However, I changed the ephemerisSettings and rotationModelSettings back to the old version instead of the interpolated spice settings. This removed the memory issue. See attached file.
`    bodySettings[ "Earth" ]->ephemerisSettings->resetFrameOrientation( "J2000" );
    bodySettings[ "Earth" ]->rotationModelSettings->resetOriginalFrame( "J2000" );
    bodySettings[ "Sun" ]->ephemerisSettings->resetFrameOrientation( "J2000" );
    bodySettings[ "Sun" ]->rotationModelSettings->resetOriginalFrame( "J2000" );`

[valgrind_non_tabulated_ephemeris.txt](https://github.com/Tudat/tudat/files/1137941/valgrind_non_tabulated_ephemeris.txt)

**Comment by DominicDirkx** [comment]
Yes, this will help, but the problem is that retrieving states directly from Spice is really quite slow...

Do you notice a serious degradation in performance?

**Comment by FrankHogervorst** [comment]
It is an increase of about 3% in CPU time. (Based on 1 measurement)

**Comment by DominicDirkx** [comment]
Huh, that's surprising (but good!). What do you use the ephemerides for? 

You're not using spherical harmonics are you? This can often become the dominant comnponent in terms of runtime.

**Comment by FrankHogervorst** [comment]
No I don't use spherical harmonics at the moment. I use it to compute whether the satellite is in shadow or not which turns the engine off and on.

**Comment by FrankHogervorst** [comment]
After I re-installed Ubuntu, I ran valgrind on the application. Now I don't get the memory leaks anymore (see attached output), even without the clearBodyMapMemory functions... Very strange. 

[valgrind_pagmo_fullLeakCheck_2017-7-14.txt](https://github.com/Tudat/tudat/files/1148027/valgrind_pagmo_fullLeakCheck_2017-7-14.txt)
